### ISSUE TITLE
Invalid yaml when decoding then encoding yaml with comments

### ISSUE THREAD

**Issue by mikefarah** [go-yaml/yaml#613]
When decoding a yaml file like:

```
rules:
- # comment
  apiGroups: [""]
```

go-yaml produces an invalid yaml when re-encoding:

```
rules:
  - apiGroups: # comment
[""]
```

See https://play.golang.org/p/wnFs49FLXep

**Comment by mikefarah** [comment]
Related: https://github.com/mikefarah/yq/issues/427

**Comment by niemeyer** [comment]
Thanks for the issue report and the patience, Mike. There were multiple issues handling that sort of comment, and they should all be fixed with a5ece68. This document should now roundtrip correctly:
```
foo:
  # Head comment of following map
  - # Head comment of bar
    bar: # Line comment of bar
      baz # Line comment of baz
```
